src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubRDDIterator.java

```diff
@@ -142,16 +148,26 @@ private List<ReceivedMessage> fetchAndAck() throws IOException {
         .build();
     }
 
-    PullResponse pullResponse = subscriber.pullCallable().call(pullRequest);
-    List<ReceivedMessage> receivedMessagesList = pullResponse.getReceivedMessagesList();
-    if (receivedMessagesList.isEmpty()) {
+    // From  the docs - "A pull response that comes with 0 messages must not be used as an indicator that there are no
+    // messages in the backlog. It's possible to get a response with 0 messages and have a subsequent
+    // request that returns messages." Since 0 messages do not indicate a lack of available messages, no delays are
+    // included  between retries.
+    // Messages will be read till batch duration completes, to account for messages published
+    // later in the batch. No documented quota on the number of synchronous subscriber requests and no issues observed
+    // in testing when repeated calls are made when no data is available in the subscription.
+    while (System.currentTimeMillis() < timeout) {
+      PullResponse pullResponse = subscriber.pullCallable().call(pullRequest);
+      List<ReceivedMessage> receivedMessagesList = pullResponse.getReceivedMessagesList();
+      if (receivedMessagesList.isEmpty()) {
+        continue;
```

Comment:
probably better to sleep a short time instead of busy-loop

Reply:
was looking at the client a little bit at https://github.com/googleapis/java-pubsub#pulling-messages, which seems to be pretty different than this API we're using. Have we looked at that one? Seems like it might be simpler to use here, as we wouldn't need to handle polling logic?

Reply:
Both approaches are documented here - https://cloud.google.com/pubsub/docs/pull#types_of_message_processing_modes . Preferred using the synchronous pull for controlling the number(and size) of messages that are acknowledged and pending processing.  Streaming pull is intended to decouple receiving and processing. The messages could accumulate and we need another mechanism to limit the number of messages read (may be streaming requests with smaller timeouts?).

Reply:
at a high level it seems like it would be pretty similar if the Queue were a bounded one. In any case, don't think we need to change the implementation right at this moment since it would be a larger change. Would still be good to sleep for some time though, unless that's somehow built into the pull call.

Reply:
Added sleep.